docs: add classport usage and examples (#11)
chains-project/classport: README.md

## About

**Classport** is a collection of tools that work together to enable **accurate runtime inspection of the supply chain in Maven Java programs**. 

> Classport's main objective is to make available information on the software supply chain at runtime.

Reviewer: ```suggestion
Classport's main objective is to make software supply chain information available at runtime.
```

To reach this goal it:

> * **Embedds** supply chain metadata in each class file that is to be part of the application

Reviewer: ```suggestion
* **Embeds** supply chain metadata in each class file that is to be part of the application during build time.
```

* **Monitors** the class loaded by the running application and **extracts** the embedded Software Supply Chain information from the loaded classes. 

In this way, Classport is able to accurately connect the loaded classes to their corresponding upstream dependencies and build a runtime representation of the application’s code-related supply chain.

## Structure

![classport overview](./resources/imgs/classport-overview.png)

The picture gives an overview of the structure of Classport. The structure is divided into two phases: **Embedding phase** and **Extracting Phase**.

### Embedding Phase

The inputs of this phase are the dependency JARs and the project source files.
The embedder is responsible for embedding the supply chain information in the form of Java Annotations into the class files.
This feature is implemented by the Maven plugin that

> modifies the JAR files during the build process and is contained in the `classport.commons` package.

Reviewer: ```suggestion
modifies the JAR files during the build process and is contained in the `classport.commons` module.
```

The JAR files from within there can then be included in the class path with the `-cp` flag as per usual.

For projects that get packaged into an Uber-JAR:

```console
mvn package -Dmaven.repo.local=classport-files
```

For multi-module projects, package each project separately as dependency properties
may differ (e.g. a direct dependency for one module is a transitive one for another).

### Dynamic analysis
Use the agent to detect the used classes:

```console
java -javaagent:<path-to-agent-jar> -jar <path-to-app-jar>
```

This command outputs the **runtime representation** of the software supply chain of the analysed project:
- `classport-deps-list` --> flat list of dependencies
- `classport-deps-tree` --> tree of dependnecies

### Static analysis

This is the command to perform static analysis:
```console
java -jar <path-to-analyzer-jar> -<printList|printTree|generateTestJar> <jarFile> [classes-to-be-ignored]
```

The available flags are:
* `-printList`, for statically generating a flat list of dependencies.

> * `-printTree`, for staticaly generating a dependencies tree.

Reviewer: Can you add example output here? It can be for the demo project that you were trying.

Use the agent to detect the used classes:

```console
java -javaagent:<path-to-agent-jar> -jar <path-to-app-jar>
```

This command outputs the **runtime representation** of the software supply chain of the analysed project:
- `classport-deps-list` --> flat list of dependencies
- `classport-deps-tree` --> tree of dependnecies

### Static analysis

This is the command to perform static analysis:
```console
java -jar <path-to-analyzer-jar> -<printList|printTree|generateTestJar> <jarFile> [classes-to-be-ignored]
```

The available flags are:
* `-printList`, for statically generating a flat list of dependencies. Below is an example of the command output on a simple Java application with only one dependency.

```
[Warning] 1 class file(s) detected without annotation for package org.example
org.apache.commons:commons-lang3:jar:3.17.0
```
* `-printTree`, for staticaly generating a dependencies tree. Below is an example of the command output on a simple Java application with only one dependency.

```
[Warning] 1 class file(s) detected without annotation for package org.example
org.example:hello:jar:0.1.0 (parent-only artefact)
\- org.apache.commons:commons-lang3:jar:3.17.0

> ```

Reviewer: Were you not able to run without warnings? It is important that we detect these dependencies with annotations. If you are able to embed successfully, then you should also be able to monitor them at runtime.

Author: Yes sorry I made a mistake. Now I think it is fixed
